File: common/src/main/java/cc/turtl/chiselmon/api/PokemonEncounter.java
package cc.turtl.chiselmon.api;

import cc.turtl.chiselmon.api.predicate.PokemonEntityPredicates;
import cc.turtl.chiselmon.api.predicate.PokemonPredicates;
import com.cobblemon.mod.common.entity.pokemon.PokemonEntity;
import com.cobblemon.mod.common.pokemon.FormData;
import com.cobblemon.mod.common.pokemon.Gender;
import com.cobblemon.mod.common.pokemon.Pokemon;
import com.cobblemon.mod.common.pokemon.Species;
import net.minecraft.client.Minecraft;
import net.minecraft.client.player.LocalPlayer;
import net.minecraft.core.registries.BuiltInRegistries;
import net.minecraft.world.level.block.state.BlockState;

import java.util.UUID;

/**
 * Immutable snapshot of a single pokemon's state when first encountered.
 */
public record PokemonEncounter(
        long encounteredMs,

        // Identity
        UUID uuid,
        Species species,
        FormData form,

        // Attributes
        int level,
        Gender gender,
        float scale,

        // Flags
        boolean isShiny,
        boolean isLegendary,

        // World context
        String dimension,
        String biome,
        long dayTime,
        boolean isRaining,

        // Block context
        String blockName,

        // Snack
        boolean spawnedFromSnack,

        // Positions
        int pokemonX,
        int pokemonY,
        int pokemonZ,
        int playerX,
        int playerY,
        int playerZ
) {

    /**
     * Creates a PokemonEncounter snapshot from a live PokemonEntity.
     *
     * @param pe The PokemonEntity to capture
     * @return A new immutable encounter record
     */
    public static PokemonEncounter from(PokemonEntity pe) {
        Pokemon pokemon = pe.getPokemon();
        LocalPlayer player = Minecraft.getInstance().player;
        BlockState blockState = pe.getBlockStateOn();

        return new PokemonEncounter(
                System.currentTimeMillis(),

                // Identity
                pe.getUUID(),
                pokemon.getSpecies(),
                pe.getForm(),

                // Attributes
                pokemon.getLevel(),
                pokemon.getGender(),
                pokemon.getScaleModifier(),

                // Flags
                PokemonPredicates.IS_SHINY.test(pokemon),
                PokemonPredicates.IS_LEGENDARY.test(pokemon),

                // World context
                pe.level().dimension().location().toString().intern(),
                pe.level().getBiome(pe.blockPosition()).getRegisteredName().intern(),
                pe.level().getDayTime() % 24000,
                pe.level().isRaining(),
                BuiltInRegistries.BLOCK.getKey(blockState.getBlock()).toString(),

                // Snack
                PokemonEntityPredicates.FROM_POKESNACK.test(pe),

                // Positions
                pe.getBlockX(),
                pe.getBlockY(),
                pe.getBlockZ(),
                player != null ? player.getBlockX() : 0,
                player != null ? player.getBlockY() : 0,
                player != null ? player.getBlockZ() : 0
        );
    }
}

File: common/src/main/kotlin/cc/turtl/chiselmon/core/api/PokemonEncounter.kt
package cc.turtl.chiselmon.core.api

import cc.turtl.chiselmon.core.api.predicate.FROM_POKESNACK
import cc.turtl.chiselmon.core.api.predicate.IS_LEGENDARY
import cc.turtl.chiselmon.core.api.predicate.IS_SHINY
import com.cobblemon.mod.common.entity.pokemon.PokemonEntity
import com.cobblemon.mod.common.pokemon.FormData
import com.cobblemon.mod.common.pokemon.Gender
import com.cobblemon.mod.common.pokemon.Species
import net.minecraft.client.Minecraft
import net.minecraft.core.registries.BuiltInRegistries
import java.util.*

/**
 * Immutable snapshot of a single pokemon's state when first encountered.
 */
data class PokemonEncounter(
    val encounteredMs: Long,

    // Identity
    val uuid: UUID,
    val species: Species,
    val form: FormData,

    // Attributes
    val level: Int,
    val gender: Gender,
    val scale: Float,

    // Flags
    val isShiny: Boolean,
    val isLegendary: Boolean,

    // World context
    val dimension: String,
    val biome: String,
    val dayTime: Long,
    val isRaining: Boolean,
    val blockName: String,

    // Snack
    val spawnedFromSnack: Boolean,

    // Positions
    val pokemonX: Int,
    val pokemonY: Int,
    val pokemonZ: Int,
    val playerX: Int,
    val playerY: Int,
    val playerZ: Int
) {
    companion object {

        /**
         * Creates a PokemonEncounter snapshot from a live PokemonEntity.
         */
        @JvmStatic
        fun from(pe: PokemonEntity): PokemonEncounter {
            val pokemon = pe.pokemon
            val player = Minecraft.getInstance().player
            val blockState = pe.blockStateOn

            val playerX = player?.blockX ?: 0
            val playerY = player?.blockY ?: 0
            val playerZ = player?.blockZ ?: 0

            return PokemonEncounter(
                encounteredMs = System.currentTimeMillis(),

                // Identity
                uuid = pe.uuid,
                species = pokemon.species,
                form = pe.form,

                // Attributes
                level = pokemon.level,
                gender = pokemon.gender,
                scale = pokemon.scaleModifier,

                // Flags
                isShiny = IS_SHINY.test(pokemon),
                isLegendary = IS_LEGENDARY.test(pokemon),

                // World context
                dimension = pe.level().dimension().location().toString().intern(),
                biome = pe.level().getBiome(pe.blockPosition()).registeredName.intern(),
                dayTime = pe.level().dayTime % 24000,
                isRaining = pe.level().isRaining,
                blockName = BuiltInRegistries.BLOCK.getKey(blockState.block).toString(),

                // Snack
                spawnedFromSnack = FROM_POKESNACK.test(pe),

                // Positions
                pokemonX = pe.blockX,
                pokemonY = pe.blockY,
                pokemonZ = pe.blockZ,
                playerX = playerX,
                playerY = playerY,
                playerZ = playerZ
            )
        }
    }
}